### PR TITLE
feat(22556): Add column mapping to the batch subscription flow

### PR DESCRIPTION
**Pull request by vanch3d**
See https://hivemq.kanbanize.com/ctrl_board/57/cards/22556/details/

This PR is the third step in the Batch Subscription epic and allows users to map columns of the data they uploaded to specific subscription properties.

The PR extracts the required properties of the subscription `JSONSchema` and the column of the worksheet, to present them in a select-based form. A simple auto-mapping (based on a Levenshtein distance) is also implemented to populate the mapping.

### Out-of-scope
- There is no type-checking on the selected mappings. It should be done in the final step of the process (validation). 

### After
screenshot-localhost_3000-2024 05 30-07_37_01

screenshot-localhost_3000-2024 05 30-07_37_14

screenshot-localhost_3000-2024 05 30-07_37_34

### CODE REVIEW

File: hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/BatchUploadButton.tsx [hunk 1]
```diff
@@ -42,11 +42,16 @@ const BatchUploadButton: FC<BatchUploadButtonProps> = () => {
         scrollBehavior="inside"
       >
         <ModalOverlay />
-        <ModalContent>
+        <ModalContent
+          style={
+            // TODO[modal-portal] Feels like a hack. Check for side-effects
+            { position: 'static' }
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/BatchUploadButton.tsx:48]: question(non-blocking): I do not know what this line does but it feel certainly like a hack. I you like we can look into that one together. But if it does the trick it fine to me.

File: hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/components/ColumnMatcherStep.tsx [hunk 2]
```diff
@@ -1,7 +1,128 @@
...
+    const { required, properties } = schema.items as JSONSchema7
+
+    //  TODO[NVL] If not required properties, do we "force" some (assuming a mqtt-topic for example)
+    return required
+      ? required.map((e) => {
+          const sss = properties?.[e] as JSONSchema7 | undefined
+          return { value: e, label: sss?.title || e }
+        })
+      : []
+  }, [schema.items])
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/components/ColumnMatcherStep.tsx:26]: suggestion: I spotted some variables that might require some love in the longrun

```suggestion
    //  TODO[NVL] If not required properties, do we "force" some (assuming a mqtt-topic for example)
    return required
      ? required.map((requiredItem) => {
          const requiredItemProperties = properties?.[requiredItem] as JSONSchema7 | undefined
          return { value: requiredItem, label: requiredItemProperties?.title || requiredItem }
        })
      : []
  }, [schema.items])
```

File: hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/components/ColumnMatcherStep.tsx [hunk 3]
```diff
@@ -1,7 +1,128 @@
...
+                        <Select<ColumnOption>
+                          // TODO[modal-portal] Feels like a hack. Use a ref [?]
+                          menuPortalTarget={document.getElementById('chakra-modal-array-field-batch') as HTMLElement}
+                          id={`mapping.${index}.column`}
+                          instanceId={`mapping.${index}.column`}
+                          aria-label={t('rjsf.batchUpload.columnMapping.column.ariaLabel')}
+                          {...rest}
+                          onChange={(e) => onChange(e?.label)}
+                          value={{ label: value, value }}
+                          options={columns}
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/components/ColumnMatcherStep.tsx:99]: issue(non-blocking): `rest` might overwrite your manually defined properties

```suggestion
                          {...rest}
                          menuPortalTarget={document.getElementById('chakra-modal-array-field-batch') as HTMLElement}
                          id={`mapping.${index}.column`}
                          instanceId={`mapping.${index}.column`}
                          aria-label={t('rjsf.batchUpload.columnMapping.column.ariaLabel')}
                          onChange={(e) => onChange(e?.label)}
                          value={{ label: value, value }}
                          options={columns}
```

File: hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/components/ColumnMatcherStep.tsx [hunk 4]
```diff
@@ -1,7 +1,128 @@
...
+                </Box>
+                <LuChevronsRight />
+                <Box flex={1}>
+                  <Input
+                    aria-label={t('rjsf.batchUpload.columnMapping.subscription.ariaLabel')}
+                    readOnly
+                    {...register(`mapping.${index}.subscription`, {
+                      required: true,
+                    })}
+                    className={errors?.mapping?.[index]?.subscription ? 'error' : ''}
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/components/ColumnMatcherStep.tsx:113]: issue(non-blocking): register properties might overwrite your manual aria-label and readOnly properties

```suggestion
                    {...register(`mapping.${index}.subscription`, {
                      required: true,
                    })}
                    aria-label={t('rjsf.batchUpload.columnMapping.subscription.ariaLabel')}
                    readOnly
                    className={errors?.mapping?.[index]?.subscription ? 'error' : ''}
```

**Review by h2xd — APPROVED** [review]
LGTM! 🐊

File: hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/BatchUploadButton.tsx (hunk 1 above)
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/BatchUploadButton.tsx:48]: The `Select` is mangled if it drops within the container of the modal.

 Screenshot 2024-05-30 at 10 22 17

There is a workaround, using the `menuPortalTarget` to point to the modal element. 
However, due to the modal being positioned `relative`, the dropdown is shifted around. 

Screenshot 2024-05-30 at 10 22 34


`static` restores the positioning, while no visible effect on the modal (inherited property I would guess)

Screenshot 2024-05-30 at 10 22 49

**Pull request merged**